src/tests/utests/iotc_utest_bsp_crypto_ecc.cc

```diff
@@ -0,0 +1,259 @@
+-----END EC PRIVATE KEY-----";
+
+class IotcBspCryptoEcc : public IotcHeapCheckTest {
+public:
+  IotcBspCryptoEcc() {
+    iotc_initialize();
+    DefaultPrivateKey.private_key_signature_algorithm =
+        IOTC_JWT_PRIVATE_KEY_SIGNATURE_ALGORITHM_ES256,
+    DefaultPrivateKey.private_key_union_type = IOTC_CRYPTO_KEY_UNION_TYPE_PEM,
+    DefaultPrivateKey.private_key_union.key_pem.key =
+        const_cast<char *>(kPrivateKey);
+  }
+  ~IotcBspCryptoEcc() { iotc_shutdown(); }
+  static constexpr uint8_t *kDefaultDataToSign =
+      (uint8_t *)"this text is ecc signed";
+  static constexpr size_t kDefaultDataToSignLength =
+      strlen((char *)kDefaultDataToSign);
+
+protected:
+  iotc_crypto_private_key_data_t DefaultPrivateKey;
+  static int ec_verify_openssl(const uint8_t *hash, size_t hash_len,
+                               const uint8_t *sig, size_t sig_len,
+                               const char *pub_key_pem);
+};
+
+/**
+ * Verify an EC signature using OpenSSL.
+ *
+ * @param hash the hash that was signed.
+ * @param hash_len length of the hash in bytes.
+ * @param sig signature to verify, formated according to rfc7518 section 3.4.
+ * @param sig_len length of the signature in bytes.
+ * @param pub_key_pem public key in PEM format.
+ * @returns 0 if the verification succeeds.
+ */
+int IotcBspCryptoEcc ::ec_verify_openssl(const uint8_t *hash, size_t hash_len,
+                                         const uint8_t *sig, size_t sig_len,
+                                         const char *pub_key_pem) {
+  if (sig_len != 64) {
+    iotc_debug_format("sig_len expected to be 64, was %d", sig_len);
+    return -1;
+  }
+
+  int ret = -1;
+  EC_KEY *ec_key_openssl = nullptr;
+  ECDSA_SIG *sig_openssl = nullptr;
+
+  // Parse the signature to OpenSSL representation
+  constexpr int kIntLength = 32; // as per RFC7518 section-3.4
+  BIGNUM *r = BN_bin2bn(sig, kIntLength, nullptr);
+  BIGNUM *s = BN_bin2bn(sig + kIntLength, kIntLength, nullptr);
+  sig_openssl = ECDSA_SIG_new();
+
+  // Parse the public key to OpenSSL representation
+  BIO *pub_key_pem_bio = BIO_new(BIO_s_mem());
+  BIO_puts(pub_key_pem_bio, pub_key_pem);
+  ec_key_openssl =
+      PEM_read_bio_EC_PUBKEY(pub_key_pem_bio, nullptr, nullptr, nullptr);
+  if (ec_key_openssl == nullptr) {
+    goto cleanup;
+  }
+
+// 1.1.0+, see opensslv.h for versions
+#if 0x10100000L <= OPENSSL_VERSION_NUMBER
+  if (1 != ECDSA_SIG_set0(sig_openssl, r, s)) {
+    goto cleanup;
+  }
+#else // Travis CI's 14.04 trusty supports openssl 1.0.1f
+  sig_openssl->r = r;
+  sig_openssl->s = s;
+#endif
+
+  if (1 == ECDSA_do_verify((const unsigned char *)hash, hash_len, sig_openssl,
+                           ec_key_openssl)) {
+    ret = 0; // Verify success
+  }
+
+cleanup:
+  ERR_print_errors_fp(stderr);
+  EC_KEY_free(ec_key_openssl);
+  ECDSA_SIG_free(sig_openssl);
+  BIO_free_all(pub_key_pem_bio);
+
+  return ret;
+}
+
+TEST_F(IotcBspCryptoEcc, SmallBuffer) {
```

Comment:
s/SmallBuffer/ReturnsErrorWhenBufferIsTooSmall

In general, you should describe the test scenario in the name of the test, so when it fails, we immediately have a good understanding, what business logic is broken.

Apply throughout the PR.